src/transaction/flashback.c

```diff
@@ -64,9 +64,99 @@ static volatile int flashback_Threshold_to_remove_archive = 0;	/* If the differe
 								 * the archive log volume can be deleted.
 								 */
 // *INDENT-OFF*
-static std::atomic_bool flashback_Is_active = false;	// the status value that the flashback is processing the request
+static std::atomic_bool flashback_Is_active = false;	// the status value that the flashback is processing the a request
+static std::atomic_bool flashback_In_progress = false;  // the status value that the flashback is in progress
 // *INDENT-ON*
 
+static CSS_CONN_ENTRY *flashback_Current_conn = NULL;	// the connection entry for a flashback requests
+
+static pthread_mutex_t flashback_Conn_lock = PTHREAD_MUTEX_INITIALIZER;
+
+/*
+ * flashback_is_duplicated_request - check if the caller is duplicated request for flashback
+ *
+ * return   : duplicated or not
+ */
+
+static bool
+flashback_is_duplicated_request (THREAD_ENTRY * thread_p)
+{
+  CSS_CONN_ENTRY *previous_conn = NULL;
+
+  /* When flashback is first called or exited properly,
+   * flashback_Current_conn is set to NULL */
+  if (flashback_Current_conn == NULL)
+    {
+      return false;
+    }
+
+  /* flashback_Current_conn indicates conn_entry in thread_p, and conn_entry can be reused by request handler.
+   * So, status in flashback_Current_conn can be overwritten. */
+
+  previous_conn = flashback_Current_conn;
+
+  if (flashback_In_progress == false)
+    {
+      /* previous flashback set flashback_In_progress to false prolperly. (exited well) */
+      return false;
+    }
+  else
+    {
+      /* flashback_In_progress == true */
+
+      if (previous_conn->in_flashback == true)
+	{
+	  /* previous flashback is still in progress */
+	  return true;
+	}
+      else
+	{
+	  /* - previous_conn is overwritten with new connection, so in_flashback value is initialized to false.
+	   * - previous flashback connection has been exited abnormally,
+	   *   so existing flashback_In_progress value has not been initialized */
+	  return false;
+	}
+    }
+}
+
+/*
+ * flashback_initialize - check request if is duplicated, 
+ *                        then initialize variables and connection when flashback request is started
+ */
+
+int
+flashback_initialize (THREAD_ENTRY * thread_p)
+{
+  /* If multiple requests come at the same time,
+   * they all can be treated as non-duplicate requests.
+   * So, latch for check and set flashback connection is required */
+
+  pthread_mutex_lock (&flashback_Conn_lock);
+
+  if (flashback_is_duplicated_request (thread_p))
+    {
+      pthread_mutex_unlock (&flashback_Conn_lock);
+
+      /* er_set() */
+      return ER_FLASHBACK_DUPLICATED_REQUEST;
+    }
+
+  flashback_Current_conn = thread_p->conn_entry;
```

Comment:
```suggestion
  assert (flashback_In_progress == false && flashback_Current_conn == NULL);

  flashback_Current_conn = thread_p->conn_entry;
```

Reply:
there is a case that previous flashback request has been shut down abnormally.
if it is shutdown abnormally, then the condition is as below
```
flashback_In_progress == true && 
flashback_Current_conn != NULL && 
flashback_Current_conn->in_flashback == false
```

So the above assert cannot be appended.

Reply:
If you found the situation that the previous flashback request was shut down abnormally, then you need to reset the global variables correctly. isn't it?

Reply:
Yes, you are right. 
I will reset the variables.